From f2a79490c1decbdfa2242e3c66550ec256b37895 Mon Sep 17 00:00:00 2001
From: Peter Swain <pswain@marvell.com>
Date: Thu, 16 May 2019 15:50:39 -0700
Subject: [PATCH 0418/1921] mmc: cavium: correct clock divisor

Rounding of the mmc clock divisor was incorrect,
allowing overclocking. We still keep clk_lo/_hi identical,
using (n, n+1) for intermediate frequencies is unproven.

Now explicitly check for overclocking.

Adjusting system clock, from which mmc clock is derived,
may allow higher frequencies when the next-smaller
divisor would take mmc clock past device maximum

Change-Id: I464fe91afa94a202e9db2682f52f29dfe1fa9aa5
Signed-off-by: Peter Swain <pswain@marvell.com>
Signed-off-by: Sujeet Baranwal <sbaranwal@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/8551
Tested-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
Reviewed-by: Chandrakala Chavva <cchavva@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/kernel/linux/+/26913
Reviewed-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
[WK: The original patch got from Marvell sdk11.21.09]
Signed-off-by: Wenlin Kang <wenlin.kang@windriver.com>
---
 drivers/mmc/host/cavium.c | 13 +++++++++++--
 1 file changed, 11 insertions(+), 2 deletions(-)

diff --git a/drivers/mmc/host/cavium.c b/drivers/mmc/host/cavium.c
index 1081057dc97e..3e0f79b0024a 100644
--- a/drivers/mmc/host/cavium.c
+++ b/drivers/mmc/host/cavium.c
@@ -1326,8 +1326,17 @@ static void cvm_mmc_set_ios(struct mmc_host *mmc, struct mmc_ios *ios)
 
 	slot->clock = clock;
 
-	if (clock)
-		clk_period = (host->sys_freq + clock - 1) / (2 * clock);
+	if (clock) {
+		clk_period = host->sys_freq / (2 * clock);
+		/* check to not exceed requested speed */
+		while (1) {
+			int hz = host->sys_freq / (2 * clk_period);
+
+			if (hz <= clock)
+				break;
+			clk_period++;
+		}
+	}
 
 	emm_switch =
 		     FIELD_PREP(MIO_EMM_SWITCH_BUS_WIDTH, bus_width) |
-- 
2.31.1

